[Velocity Tracker] Fix: Issue 97761: Flutter Scrolling does not match iOS; inadvertent scrolling when user lifts up finger

chipweinberger opened the pull request:
## Issue

**Issue:** https://github.com/flutter/flutter/issues/97761

https://github.com/flutter/flutter/assets/1863934/53c5e0df-b85a-483c-a17d-bddd18db3aa9

## The Cause:

The bug is very simple to understand - `velocity_tracker.dart` **only adds new samples while your finger is moving**.

**Therefore**, if you move your finger quickly & (important) stop suddenly with no extra movement, the last 3 samples will all be > 0 dy. Regardless of how long you wait, you will get movement when you lift up your finger.

**Logs from velocity_tracker.dart:**
Notice: all 3 `_previousVelocityAt` are `dy > 0` despite a 2 second delay since the last scroll
```
// start moving finger
flutter: addPosition dy:-464.0
flutter: addPosition dy:-465.0
flutter: addPosition dy:-466.0
flutter: addPosition dy:-467.0
flutter: addPosition dy:-468.0
flutter: addPosition dy:-469.0
flutter: addPosition dy:-470.0
// stop moving finger here, keep it still for 2 seconds & lift it up
flutter: _previousVelocityAt(-2) samples(-467.0, -468.0)) dy:-176.772140710624
flutter: _previousVelocityAt(-1) samples(-468.0, -469.0)) dy:-375.0937734433609
flutter: _previousVelocityAt(0) samples(-469.0, -470.0)) dy:-175.71604287471447
flutter: primaryVelocity DragEndDetails(Velocity(0.0, -305.5)).primaryVelocity
flutter: createBallisticSimulation pixels 464.16666666666663 velocity 305.4699824197211
```

## The Fix

**There are 3 options to fix it:**
A. sample uniformly *per unit time* (a larger more risky change, hurts battery life)
B. consider elapsed time since the last sample. If greater than X, assume no more velocity. (easy & just as valid)
C. similar to B, but instead add "ghost samples" of velocity zero, and run calculations as normal (a bit tricker, of dubious benefit imo)

**For Option B I considered two approaches:**
1. _get the current timestamp and compare to event timestamp._  This is tricky because events are documented to use an arbitrary timescale & I wasn't able to find the code that generates the timestamps. This approach could be considered more.
2. _get a new timestamp using Stopwatch and compare now vs when the last sample was added._ This is the solution implemented here.  There is a limitation in that we don't know when addSamples is called relative to the event. But, this estimation is already on a very low latency path & still it gives us a *minimum* time bound which is sufficient for comparison. 

**This PR chooses the simplest of the all solutions (B2). Please try it our yourself, it completely solves the problem 😀** Option _B1_ would be a nice alternative as well, if we can define and access the same timesource as the pointer tracker in a maintainable simple way.

## After Fix

https://github.com/flutter/flutter/assets/1863934/be50d8e7-d5da-495a-a4af-c71bc541cbe3



## Pre-launch Checklist
- [X] I signed the [CLA].

chipweinberger committed:
[Velocity Tracker] Fix: Issue 97761: Flutter Scrolling does not match iOS; inadvertent scrolling when user lifts up finger

chipweinberger commented:
@goderbauer @Piinks 

Thanks for your time.

Piinks reviewed (commented):
Hi @chipweinberger welcome, thank you for contributing. This will need tests to verify that it works as expected. Also, I noticed the video captures the bug, can you share one that illustrates the fix?

We often use the scroll_overlay app in flutter/platform_tests to see flutter scrolling laid on top of native scrolling in order to compare. That may be helpful if you are interested in using the tool to check the behavior. :)

Piinks commented:
It also looks like many tests are failing here. Can you take a look?

chipweinberger commented:
@Piinks of course. 

I have added a simple test & a video of the fix. 

The failing tests failing look unrelated to this PR. Weird build related error messages. I have updated the branch, hopefully someone else has fixed it.

`Working directory: /Volumes/Work/s/w/ir/x/w/flutter/dev/integration_tests/spell_check\nstdout and stderr output:\n(stdout/stderr output was more than 10 lines)\n  ð���  ð���  \nCommand: ../../../bin/flutter build apk --release -v\nCommand exited with exit code 1 but expected zero exit code.\nWorking directory: /Volumes/Work/s/w/ir/x/w/flutter/dev/integration_tests/spell_check\nstdout and stderr `

Piinks reviewed (commented):
It looks like there are still a lot of failing tests here, can you take a look?

chipweinberger commented:
I've looked and rebased a few times.

Unfortunately it's very outside of my knowledge area.

Can someone lend some insight?

Piinks commented:
The checks are failing because of this change, so rebasing will not resolve them. It looks like it primarily has to do with the analyzer. Have you run `flutter analyze --flutter-repo`?

chipweinberger commented:
thanks for the suggestion. It spits out many hundreds or errors. None of them seem to make actual sense. Here's some of the few that referenced `velocity_tracker.dart`

```
 Target of URI doesn't exist: 'package:flutter/foundation.dart' •
         ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:6:8 • uri_does_not_exist
  error • Target of URI doesn't exist: 'dart:ui' • ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:11:8 •
         uri_does_not_exist
  error • Undefined name 'immutable' used as an annotation •
         ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:14:1 • undefined_annotation
   info • Method invocation or property access on a 'dynamic' target •
          ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:24:58 • avoid_dynamic_calls
  error • Const variables must be initialized with a constant value •
         ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:24:58 •
         const_initialized_with_non_constant_value
  error • Invalid constant value • ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:24:58 •
         invalid_constant
  error • Undefined name 'Offset' • ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:24:58 •
         undefined_identifier
  error • Undefined class 'Offset' • ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:27:9 •
         undefined_class
   info • Method invocation or property access on a 'dynamic' target •
          ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:30:55 • avoid_dynamic_calls
   info • Method invocation or property access on a 'dynamic' target •
          ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:34:38 • avoid_dynamic_calls
   info • Method invocation or property access on a 'dynamic' target •
          ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:39:38 • avoid_dynamic_calls
   info • Method invocation or property access on a 'dynamic' target •
          ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/packages/flutter/lib/src/gestures/velocity_tracker.dart:55:33 • avoid_dynamic_calls
  error • A value of type 'dynamic' can't be assigned to a variable of type 'double' •
         ../../../../Volumes/User/MBP-Google-Drive/jamcorder/app/flutter/pac
```

Piinks commented:
> lib/src/gestures/velocity_tracker.dart:351:47: Error: The getter '_assumePointerMoveStoppedMilliseconds' isn't defined for the class 'IOSScrollViewFlingVelocityTracker'.
 - 'IOSScrollViewFlingVelocityTracker' is from 'package:flutter/src/gestures/velocity_tracker.dart' ('lib/src/gestures/velocity_tracker.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named '_assumePointerMoveStoppedMilliseconds'.
    if (sinceLastSample.elapsedMilliseconds > _assumePointerMoveStoppedMilliseconds) {

It looks like the issue is you are accessing a static member in correctly. `_assumePointerMoveStoppedMilliseconds` should be `VelocityTracker._assumePointerMoveStoppedMilliseconds`.

chipweinberger commented:
Thanks. I've updated the PR. 

To ease any testing worry, this change *has* been tested and does work. Using that constant was a last second change. 

Thanks again.

Piinks commented:
Thanks! I am not worried. That's why we have tests. If there is an issue, they fail. :)

Piinks commented:
Looks like we're still not passing CI here, I've rebased to see if it resolves the failures.

goderbauer commented on packages/flutter/test/gestures/velocity_tracker_test.dart:
```diff
@@ -144,4 +144,22 @@ void main() {
     }
   });
+
+  test('Assume zero velocity when there are no recent samples', () {
```
This method needs to be async, which may fix all/most of the CI errors.

```suggestion
  test('Assume zero velocity when there are no recent samples', () async {
```

chipweinberger replied:
thanks for taking a look at this! 

i'll push an update! (later tonight)

Piinks reviewed (commented):
Unfortunately it looks like checks are still failing here. Can you take a look?

```
   info • Missing documentation for a public member • packages/flutter/lib/src/gestures/velocity_tracker.dart:157:19 • public_member_api_docs
   info • Missing type annotation • packages/flutter/test/gestures/velocity_tracker_test.dart:161:11 • always_specify_types
   info • Use 'const' with the constructor to improve performance • packages/flutter/test/gestures/velocity_tracker_test.dart:161:26 • prefer_const_constructors
   info • Use 'const' with the constructor to improve performance • packages/flutter/test/gestures/velocity_tracker_test.dart:163:51 • prefer_const_constructors
```

I highly recommend running `flutter analyze --flutter-repo` as well as `flutter test` locally to ensure everything passes.

chipweinberger commented:
thanks @Piinks.

i'm away for a week, but will after :)

chipweinberger commented:
ran it locally, looks good

```
chipweinberger@Chips-MacBook-Pro flutter % /Users/chipweinberger/Downloads/flutter/bin/flutter test --name "Assume zero velocity when there are no recent samples"
02:37 +1: All tests passed! 
```

Piinks reviewed (approved):
LGTM, thank you for resolving the failing checks!
